### PR TITLE
Add TO_TEXT_CONTRACT_ID primitive

### PR DESCRIPTION
**Pull request by cocreature**
This is the first part of #7114.

This PR

* Adds the primitive to the protobuf.
* Handles decoding and encoding in Haskell and Scala.
* Handles typechecking in Haskell and Scala.
* Handles speedy compilation and interpretation in Scala.
* Updates the specification.

This PR does not yet change the standard library to make use of this
primitive.

changelog_begin
changelog_end

### Pull Request Checklist

- [ ] Read and understand the [contribution guidelines](https://github.com/digital-asset/daml/blob/master/CONTRIBUTING.md)
- [ ] Include appropriate tests
- [ ] Set a descriptive title and thorough description
- [ ] Add a reference to the [issue this PR will solve](https://github.com/digital-asset/daml/issues), if appropriate
- [ ] Include changelog additions in one or more commit message bodies between the `CHANGELOG_BEGIN` and `CHANGELOG_END` tags
- [ ] Normal production system change, include purpose of change in description

NOTE: CI is not automatically run on non-members pull-requests for security
reasons. The reviewer will have to comment with `/AzurePipelines run` to
trigger the build.

### CODE REVIEW

File: daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala [hunk 1]
```diff
@@ -1556,6 +1556,10 @@ private[lf] object DecodeV1 {
       BuiltinFunctionInfo(TO_TEXT_TIMESTAMP, BToTextTimestamp),
       BuiltinFunctionInfo(TO_TEXT_PARTY, BToTextParty, minVersion = partyTextConversions),
       BuiltinFunctionInfo(TO_TEXT_TEXT, BToTextText),
+      BuiltinFunctionInfo(
+        TO_TEXT_CONTRACT_ID,
+        BToTextContractId,
+        minVersion = contractIdTextConversions),
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/main/scala/com/digitalasset/daml/lf/archive/DecodeV1.scala:1562]:
```suggestion
        minVersion = contractIdTextConversions,
    ),
```

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 2]
```diff
@@ -114,6 +114,16 @@ class DecodeV1Spec
...
   )
 
+  private val preContractIdTextConversionVersions = Table(
+    "minVersion",
+    List(1, 4, 6, 7).map(i => LV.Minor.Stable(i.toString)): _*
+  )
+
+  private val postContractIdTextConversionVersions = Table(
+    "minVersion",
+    LV.Minor.Dev
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:124]:
```suggestion
    // FIXME: https://github.com/digital-asset/daml/issues/7139
    // uncomment the following line once LF 1.9 is released
    // LV.Minor.Stable("9"),
    LV.Minor.Dev,
```

File: daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala [hunk 3]
```diff
@@ -114,6 +114,16 @@ class DecodeV1Spec
     LV.Minor.Dev,
   )
 
+  private val preContractIdTextConversionVersions = Table(
+    "minVersion",
+    List(1, 4, 6, 7).map(i => LV.Minor.Stable(i.toString)): _*
```
Comment (remyhaemmerle-da) [inline daml-lf/archive/src/test/scala/com/digitalasset/daml/lf/archive/DecodeV1Spec.scala:119]:
```suggestion
    List(1, 4, 6, 8).map(i => LV.Minor.Stable(i.toString)): _*
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 4]
```diff
@@ -305,6 +305,22 @@ private[lf] object SBuiltin {
...
   }
 
+  final case object SBToTextContractId extends SBuiltin(1) {
+    override private[speedy] final def execute(
+        args: util.ArrayList[SValue],
+        machine: Machine): Unit = {
+      args.get(0) match {
+        case SContractId(cid) =>
+          if (machine.onLedger) {
+            machine.returnValue = SOptional(None)
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:315]:
```suggestion
            machine.returnValue = SValue.SValue.None
```

File: daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala [hunk 5]
```diff
@@ -305,6 +305,22 @@ private[lf] object SBuiltin {
...
+          } else {
+            machine.returnValue = SOptional(Some(SText(cid.coid)))
+          }
```
Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:318]:
```suggestion
          machine.returnValue = 
              if (machine.onLedger) SValue.SValue.None else  SOptional(Some(SText(cid.coid)))
```

Comment (remyhaemmerle-da) [inline daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:318]:
```suggestion
      machine.returnValue = args.get(0) match {
        case SContractId(cid) if (machine.onLedger) =>
             SOptional(Some(SText(cid.coid)))
        case _ => 
             SValue.SValue.None
      }  
```

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:318]: This suggestion isn’t correct.
1. You flipped the flag. You are returning `Some` in the `onLedger` case.
2. I would like to validate that we get a contract id even in the onLedger case. Just makes it easier to catch bugs.

Comment (remyhaemmerle-da) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala:318]: My bad.
You could still use `SValue.SValue.None` that save you an allocation.

Comment (cocreature) [reply daml-lf/interpreter/src/main/scala/com/digitalasset/daml/lf/speedy/SBuiltin.scala]: Thanks, changed!

**Comment by remyhaemmerle-da** [comment]
As we discuss off line. 
We agree it is better for now to not touch the `fromScenarioExpr` and `fromSExpr` and postpone the change of on-ledger built-ins.

**Pull request merged**